### ISSUE TITLE
Special HID codes

### ISSUE THREAD

**Issue by driechers** [hak5darren/USB-Rubber-Ducky#48]
It would be nice to be able to send special HID events such as:

- 0x00 Reserved (no event indicated)
- 0x01 Keyboard ErrorRollOver
- 0x02 Keyboard ErrorUndefined

Proposed solution:
New command where a user could specify hid codes in hex for example, "HID 0x01"

**Comment by mame82** [comment]
I vote for this idea, but would suggest sending a keycode byte and a modifier byte, like:

`HID <key> <modifier>`

To produce a pipe '|' operator (ASCII 0x7C) on German keyboard layout the following could be used:
`HID KEY_NON_US_100 MODIFIERKEY_RIGHT_ALT`

Multiple modifiers like in CTRL+ALT+t could be done like this:
`HID KEY_T (MODIFIERKEY_CTRL | MODIFIERKEY_ALT)`

With this it should even be possible to send non-ASCII characters, like "Volume Up"
`HID 0x80 0x00`

**Comment by mame82** [comment]
@driechers @hak5darren 

I've implemented your idea in this fork [fork](https://github.com/mame82/USB-Rubber-Ducky/tree/USB-HID-keycode)

You have to compile Encoder.java by hand as I haven't rebuilt the jar archive.

An example script could be found in [HID_command_test.script](https://github.com/mame82/USB-Rubber-Ducky/blob/USB-HID-keycode/HID_command_test.script)

Please help testing. If you confirm this is working I'll initiate a pull request

**Comment by driechers** [comment]
<pre>
DELAY 3000
HID 0x80 0x00
HID 0x80 0x00
HID 0x80 0x00
HID 0x80 0x00
HID 0x80 0x00
HID 0x80 0x00
HID 0x05 0x00
</pre>
With this it's typing "hhhhhhh", when I would expect volume up 6x followed by b

**Comment by mame82** [comment]
@driechers Encoding your script with patched Encoder.java gives the following result (in hex):

```
# xxd inject.bin
00000000: 00ff 00ff 00ff 00ff 00ff 00ff 00ff 00ff  ................
00000010: 00ff 00ff 00ff 00c3 8000 8000 8000 8000  ................
00000020: 8000 8000 0500                           ......
```

* Every `0x00FF` means a 255 ms delay
* The `0x00C3` is an additional delay of 195 ms
* This sums up to exactly 3000ms delay
* This is followed by 6 times `0x8000` which translates to HID keycode 0x80 without a modifier.
* The `0x0500` translates to key 0x05 without modifier (lowercase "b" on US layout keyboard)

**The good news first:**
Printing out the "b" works as intended on a RubberDucky with stock firmware in all my tests. So this is likely to be an issue with your RuberDucky firmware or the script produces wrong output on your side (compare the hex dump shown above with your result to check this).

**Now the bad news:**
The default HID firmware only implements the HID keyboard profile (look [here](https://github.com/hak5darren/USB-Rubber-Ducky/blob/master/Firmware/Source/Ducky_HID/src/asf/common/services/usb/class/hid/device/kbd/udi_hid_kbd.c#L107) for details). This means that key codes which aren't meant to be send from a "PC-AT" keyboard aren't processed. This is basically true for all USB HID codes larger than 0x65 (see HID keyboard usage table for details  [here](http://www.usb.org/developers/hidpage/Hut1_12v2.pdf) on page 65).

In fact, to make media keys work, the firmware has to be changed to support the "Generic Consumer Device" profile and code has to be added to send the corresponding multimedia keys. An example implementation for an Arduino based device could be found [here](https://github.com/aduitsis/ardumultimedia).

**To sum up:** 
The implementation works as intended on my side, with the restriction that one is only able to send "PC-AT" HID keycodes as outlined in the USB HID keyboard usage table.

Encoding this script with the patched Encoder.java ...
```
DELAY 3000
HID 0x04 0x00
HID 0x04 0x02
HID KEY_B
HID KEY_B MODIFIERKEY_LEFT_SHIFT
HID KEY_C 0x00
HID KEY_C 0x02
```

... results in this output:
`aAbBcC`

So I consider this working, with the restriction outlined above.

> With this it should even be possible to send non-ASCII characters, like "Volume Up"

has been a bad usage example.

**Comment by mame82** [comment]
One more addition, the updated encoder wouldn't allow to send 

> 0x00 Reserved (no event indicated)

The keycode 0x00 is reserved for delay on RubberDucky, with the modifier being the delay time in milliseconds. As the DELAY feature is already implemented `HID 0x00 0xff` will be ignored by the patch.

**Comment by mame82** [comment]
Pull request to merge "HID" command into master branch [here](https://github.com/hak5darren/USB-Rubber-Ducky/pull/54)

**Comment by Fardenco** [comment]
I've found this page which shows windows media key codes, I don't really understand how it works but maybe it might help

**Comment by mame82** [comment]
@Fardenco The codes for USB keyboards are defined by the "[USB HID usage tables](http://www.usb.org/developers/hidpage/Hut1_12v2.pdf)" where table 12 describes keyboard usage on page 52. The mapping from the USB keycode to your language is done by the OS driver. For example on AWERTY keycode 0x14 (q)  should translate to "a". If the defined media keys are usable depends on the driver support. If they don't work a " Generic Consumer Device" has to be added to the table

**Comment by Fardenco** [comment]
Ok so it's only about changing the device descriptor ?
But why is the keycode 0x80 not working on Windows ? Should we look for another one ?
Maybe is there a way to sniff a standard keyboard to see what is sent whent we press the media keys ?

**Comment by mame82** [comment]
> Ok so it's only about changing the device descriptor ?
No, look at the given example. Another descriptor mostly results in different USB EPs and thus in different implementation.

My (hacky) change to the descriptor raised the max values for the EPs already present. Basically the media key 0x80 is from the same usage table, but greater than the defined MAX value, thus I changed it. As outlined in the usage table, 0x80 is a Unix key
